### PR TITLE
Pull on start test for Google Cloud: Pub/Sub

### PR DESCRIPTION
**Pull request by danimegar**
| Related issue | 
| --- |
| #618 |

## Description
I have created 2 tests: 
1. The module has to pull messages after restarting Wazuh if 'pull_on_start' parameter is set to 'yes'.
2.  The module starts to pull after the time interval when the 'pull_on_start' parameter is set to 'no'.

## Configuration
```
<gcp-pubsub>
  <pull_on_start>yes</pull_on_start>
  <interval>30s</interval>
  <project_id>sinuous-voice-271711</project_id>
  <subscription_name>wazuh-integration</subscription_name>
  <credentials_file>credentials.json</credentials_file>
</gcp-pubsub>
```
## Output test
pull_on_start

### CODE REVIEW

File: deps/wazuh_testing/wazuh_testing/gcloud.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Copyright (C) 2015-2020, Wazuh Inc.
+# Created by Wazuh, Inc. <info@wazuh.com>.
+# This program is free software; you can redistribute it and/or modify it under the terms of GPLv2
+
+import os
+
+_data_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'data')
+
+def callback_detect_start_gcp(line):
```
Comment (crd1985) [inline deps/wazuh_testing/wazuh_testing/gcloud.py:9]: PEP8. 2 blank spaces required.

File: tests/integration/test_gcloud/conftest.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+from py.xml import html
+
+from wazuh_testing import global_parameters
+from wazuh_testing.gcloud import detect_gcp_start
+from wazuh_testing.tools.configuration import get_wazuh_conf, write_wazuh_conf, set_section_wazuh_conf
+from wazuh_testing.tools import WAZUH_CONF
+from wazuh_testing.tools.services import control_service
+
+@pytest.fixture(scope='module')
+def wait_for_gcp_start(get_configuration, request):
```
Comment (crd1985) [inline tests/integration/test_gcloud/conftest.py:18]: PEP8. Two blank spaces required.

File: tests/integration/test_gcloud/conftest.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    # Wait for module gpc-pubsub starts
+    file_monitor = getattr(request.module, 'wazuh_log_monitor')
+    detect_gcp_start(file_monitor)
+
+@pytest.fixture(scope='module')
+def configure_environment(get_configuration, request):
```
Comment (crd1985) [inline tests/integration/test_gcloud/conftest.py:24]: PEP8. Two blank spaces required.

File: tests/integration/test_gcloud/conftest.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+# Copyright (C) 2015-2020, Wazuh Inc.
+# Created by Wazuh, Inc. <info@wazuh.com>.
+# This program is free software; you can redistribute it and/or modify it under the terms of GPLv2
+
+import os
```
Comment (crd1985) [inline tests/integration/test_gcloud/conftest.py:5]: Import not used.

File: tests/integration/test_gcloud/conftest.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+import pytest
+from numpydoc.docscrape import FunctionDoc
+from py.xml import html
```
Comment (crd1985) [inline tests/integration/test_gcloud/conftest.py:9]: Imports not used.

File: tests/integration/test_gcloud/conftest.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+from wazuh_testing import global_parameters
+from wazuh_testing.gcloud import detect_gcp_start
+from wazuh_testing.tools.configuration import get_wazuh_conf, write_wazuh_conf, set_section_wazuh_conf
+from wazuh_testing.tools import WAZUH_CONF
```
Comment (crd1985) [inline tests/integration/test_gcloud/conftest.py:14]: Import not used.

File: tests/integration/test_gcloud/test_pull_on_start.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+@pytest.fixture(scope='module', params=configurations)
+def get_configuration(request):
+    """Get configurations from the module."""
+    return request.param
+
+
+# tests
+
+def test_pull_on_start(get_configuration, configure_environment,
+                   restart_wazuh, wait_for_gcp_start):
```
Comment (crd1985) [inline tests/integration/test_gcloud/test_pull_on_start.py:58]: Bad indentation.

File: tests/integration/test_gcloud/test_pull_on_start.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    value = get_configuration['elements'][0]['pull_on_start']['value'] == 'yes'
+    sinterval = get_configuration['elements'][4]['interval']['value']
+    interval = int(''.join(filter(str.isdigit, sinterval)))
+
+    if value:
+        wazuh_log_monitor.start(timeout=global_parameters.default_timeout,
+                                callback=callback_detect_start_fetching_logs,
+                                accum_results=1,
+                                error_message='Did not receive expected '
+                                                   '"Starting fetching of logs" event').result()
```
Comment (crd1985) [inline tests/integration/test_gcloud/test_pull_on_start.py:73]: Bad indentation.

File: tests/integration/test_gcloud/test_pull_on_start.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    else:
+        wazuh_log_monitor.start(timeout=global_parameters.default_timeout,
+                            callback=callback_detect_start_gcp_sleep,
+                            error_message='Did not receive expected '
+                                          '"Sleeping for x seconds" event')
```
Comment (crd1985) [inline tests/integration/test_gcloud/test_pull_on_start.py:78]: Bad indentation.

File: tests/integration/test_gcloud/test_pull_on_start.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        with pytest.raises(TimeoutError):
+            event = wazuh_log_monitor.start(timeout=global_parameters.default_timeout, callback=callback_detect_start_fetching_logs)
+            raise AttributeError(f'Unexpected event {event}')
+
+        time.sleep(interval)
+        wazuh_log_monitor.start(timeout=global_parameters.default_timeout,
+                                callback=callback_detect_start_fetching_logs,
+                                accum_results=1,
+                                error_message='Did not receive expected '
+                                                   '"Starting fetching of logs" event').result()
```
Comment (crd1985) [inline tests/integration/test_gcloud/test_pull_on_start.py:88]: Bad indentation.

File: tests/integration/test_gcloud/test_pull_on_start.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    else:
+        wazuh_log_monitor.start(timeout=global_parameters.default_timeout,
+                            callback=callback_detect_start_gcp_sleep,
+                            error_message='Did not receive expected '
+                                          '"Sleeping for x seconds" event')
+        with pytest.raises(TimeoutError):
+            event = wazuh_log_monitor.start(timeout=global_parameters.default_timeout, callback=callback_detect_start_fetching_logs)
```
Comment (crd1985) [inline tests/integration/test_gcloud/test_pull_on_start.py:80]: Line too long.

File: tests/integration/test_gcloud/test_pull_on_start.py [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+# configurations
+
+monitoring_modes = ['scheduled']
+conf_params = {'PROJECT_ID': project_id, 'SUBSCRIPTION_NAME': subscription_name,
+               'CREDENTIALS_FILE': credentials_file, 'INTERVAL': interval, 
+               'MAX_MESSAGES': max_messages, 'LOGGING': logging, 'MODULE_NAME': __name__}
+
+p, m = generate_params(extra_params=conf_params, 
+                       apply_to_all=({'PULL_ON_START': pull_on_start_value} for  pull_on_start_value in pull_on_start),
```
Comment (crd1985) [inline tests/integration/test_gcloud/test_pull_on_start.py:41]: Two spaces after `for`

**Comment by crd1985** [comment]
Also be careful about the recent changes in all yaml configurations. Now configurations are allowed to refer to more than one section. Review how it impacts to yours.

File: tests/integration/test_gcloud/test_pull_on_start.py [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+# tests
+
+def test_pull_on_start(get_configuration, configure_environment,
+                   restart_wazuh, wait_for_gcp_start):
+    """
+    Verify the module gcp-pubsub starts if pull_on_start is set to yes and 
+    the module sleeps if pull_on_start is set to no.
+    In the second case, the module will start to pull messages after time interval.  
+    """
+    value = get_configuration['elements'][0]['pull_on_start']['value'] == 'yes'
```
Comment (crd1985) [inline tests/integration/test_gcloud/test_pull_on_start.py:64]: Use a more intuitive var name.

File: tests/integration/test_gcloud/test_pull_on_start.py (hunk 10 above)
Comment (crd1985) [inline tests/integration/test_gcloud/test_pull_on_start.py:88]: This test does not guarantee no events are fetch before the interval. Maybe you should use that interval as a timeout for expecting a timeout error trying to fetch events.

Comment (danimegar) [reply tests/integration/test_gcloud/test_pull_on_start.py]: Done.

File: tests/integration/test_gcloud/test_pull_on_start.py (hunk 13 above)
Comment (danimegar) [reply tests/integration/test_gcloud/test_pull_on_start.py]: Done.

File: tests/integration/test_gcloud/test_pull_on_start.py (hunk 12 above)
Comment (danimegar) [reply tests/integration/test_gcloud/test_pull_on_start.py]: Fixed.

File: tests/integration/test_gcloud/test_pull_on_start.py (hunk 11 above)
Comment (danimegar) [reply tests/integration/test_gcloud/test_pull_on_start.py]: Fixed

File: tests/integration/test_gcloud/test_pull_on_start.py (hunk 10 above)
Comment (danimegar) [reply tests/integration/test_gcloud/test_pull_on_start.py]: Fixed

File: tests/integration/test_gcloud/test_pull_on_start.py (hunk 9 above)
Comment (danimegar) [reply tests/integration/test_gcloud/test_pull_on_start.py]: Fixed

File: tests/integration/test_gcloud/test_pull_on_start.py (hunk 8 above)
Comment (danimegar) [reply tests/integration/test_gcloud/test_pull_on_start.py]: Fixed

File: tests/integration/test_gcloud/test_pull_on_start.py (hunk 7 above)
Comment (danimegar) [reply tests/integration/test_gcloud/test_pull_on_start.py]: Fixed

File: tests/integration/test_gcloud/conftest.py (hunk 6 above)
Comment (danimegar) [reply tests/integration/test_gcloud/conftest.py]: Fixed

File: tests/integration/test_gcloud/conftest.py (hunk 5 above)
Comment (danimegar) [reply tests/integration/test_gcloud/conftest.py]: Fixed

File: tests/integration/test_gcloud/conftest.py (hunk 4 above)
Comment (danimegar) [reply tests/integration/test_gcloud/conftest.py]: Fixed

File: tests/integration/test_gcloud/conftest.py (hunk 3 above)
Comment (danimegar) [reply tests/integration/test_gcloud/conftest.py:21]: Fixed

File: tests/integration/test_gcloud/conftest.py (hunk 2 above)
Comment (danimegar) [reply tests/integration/test_gcloud/conftest.py:14]: Fixed

File: deps/wazuh_testing/wazuh_testing/gcloud.py (hunk 1 above)
Comment (danimegar) [reply deps/wazuh_testing/wazuh_testing/gcloud.py:10]: Fixed

**Pull request merged**